course5/homework06/main.cc

```diff
@@ -0,0 +1,266 @@
+    assert(strncmp_result == 0);
+
+    for (auto i = 0; i < expected_length; i++)
+    {
+        assert(s[i] == expected_content[i]);
+    }
+
+    std::cout << "Passed: " << test_name << std::endl;
+}
+
+template<typename StrT, typename StrT2>
+void testEquality(const char* test_name, const StrT& s1, const StrT2& s2, bool are_equal)
+{
+    if (are_equal)
+    {
+        assert(s1 == s2);
+        assert(s2 == s1);
+    }
+    else
+    {
+        assert(s1 != s2);
+        assert(s2 != s1);
+    }
+
+    std::cout << "Passed: " << test_name << std::endl;
+}
+
+int main(int, char**)
+{
+    using namespace simple_string;
+    using namespace std::literals;
+
+    /* Basic tests for construction and assignment */
+    String defaultS;
+    String sFromChar = "word";
+    String copyConstructed = sFromChar;
+
+    String toBeAssigned = "movetest";
+    String moveConstructed = std::move(toBeAssigned);
+
+    toBeAssigned = "assignment";
+    String copyAssigner = "CopyAssignER";
+    String moveAssigner = "MoveAssignER";
+    String copyAssigned;
+    copyAssigned = copyAssigner;
+    String moveAssigned;
+    moveAssigned = std::move(moveAssigner);
+
+    copyAssigner[2] = 'z';
+
+    const int num_test_cases = 8;
+    std::array<const char *, num_test_cases> test_names = {
+        {
+            "default constructor",
+
+            "const char * construction",
+            "copy construction",
+            "move construction",
+
+            "const char * assignment",
+            "copy assignment",
+            "move assignment",
+
+            "[] assignment"
+        }
+    };
+    std::array<String, num_test_cases> inputs = {
+        {
+            defaultS,
+            sFromChar, copyConstructed, moveConstructed,
+            toBeAssigned, copyAssigned, moveAssigned,
+            copyAssigner
+        }
+    };
+    std::array<const char *, num_test_cases> expected_contents = {
+        {
+            "",
+            "word", "word", "movetest",
+            "assignment", "CopyAssignER", "MoveAssignER",
+            "CozyAssignER",
+        }
+    };
+
+    for (auto i = 0; i < num_test_cases; i++)
+    {
+        test(test_names[i], inputs[i], expected_contents[i]);
+    }
+
+    /* Tests for concatenation */
+    String concatStr = "Hello ";
+    String concatToConstChar = concatStr + "World!";
+    test("concat initialization w const char *", concatToConstChar, "Hello World!");
+
+    String concatStr2 = "World Again!";
+    String concatToString = concatStr + concatStr2;
+    test("concat initialization w String", concatToString, "Hello World Again!");
+
+    concatToString += " And again.";
+    test("+= w const char *", concatToString, "Hello World Again! And again.");
+
+    String concatStr3 = " And again twice.";
+    concatToString += concatStr3;
+    test("+= w String", concatToString, "Hello World Again! And again. And again twice.");
+
+    /* Test out of bounds error with indexing */
```

Comment:
I do have a `catch` block. I think it's something about the index-assignment out of bounds?
```
==1508== HEAP SUMMARY:
==1508==     in use at exit: 26,925 bytes in 193 blocks
==1508==   total heap usage: 308 allocs, 115 frees, 33,706 bytes allocated
==1508== 
==1508== 16 bytes in 1 blocks are definitely lost in loss record 4 of 65
==1508==    at 0x1000175B9: calloc (in /usr/local/Cellar/valgrind/3.11.0/lib/valgrind/vgpreload_memcheck-amd64-darwin.so)
==1508==    by 0x1000F8636: __cxa_get_globals (in /usr/lib/libc++abi.dylib)
==1508==    by 0x1000F8A19: __cxa_throw (in /usr/lib/libc++abi.dylib)
==1508==    by 0x1000052DE: simple_string::String::operator[](int) (simple_string.cc:181)
==1508==    by 0x1000069A2: main (main.cc:150)
==1508== 
==1508== 272 bytes in 1 blocks are definitely lost in loss record 50 of 65
==1508==    at 0x100016EBB: malloc (in /usr/local/Cellar/valgrind/3.11.0/lib/valgrind/vgpreload_memcheck-amd64-darwin.so)
==1508==    by 0x10023EB58: getPerThreadBufferFor_dlerror(unsigned long) (in /usr/lib/system/libdyld.dylib)
==1508==    by 0x7FFF5FC0C0C4: dlerrorSet(char const*) (in /usr/lib/dyld)
==1508==    by 0x7FFF5FC0C3F2: dlopen (in /usr/lib/dyld)
==1508==    by 0x10023E79B: dlopen (in /usr/lib/system/libdyld.dylib)
==1508==    by 0x1000FBC40: std::__1::__libcpp_refstring::compute_gcc_empty_string_storage() (in /usr/lib/libc++abi.dylib)
==1508==    by 0x1000FBC0E: std::__1::__libcpp_refstring::get_gcc_empty_string_storage() (in /usr/lib/libc++abi.dylib)
==1508==    by 0x1000FBBB1: std::__1::__libcpp_refstring::~__libcpp_refstring() (in /usr/lib/libc++abi.dylib)
==1508==    by 0x1000FBA51: std::range_error::~range_error() (in /usr/lib/libc++abi.dylib)
==1508==    by 0x1000F8BD2: __cxa_decrement_exception_refcount (in /usr/lib/libc++abi.dylib)
==1508==    by 0x100006A6D: main (main.cc:156)
==1508== 
==1508== 2,064 bytes in 1 blocks are possibly lost in loss record 59 of 65
==1508==    at 0x10001717C: malloc_zone_malloc (in /usr/local/Cellar/valgrind/3.11.0/lib/valgrind/vgpreload_memcheck-amd64-darwin.so)
==1508==    by 0x1005F0EFD: _objc_copyClassNamesForImage (in /usr/lib/libobjc.A.dylib)
==1508==    by 0x1005E4182: protocols() (in /usr/lib/libobjc.A.dylib)
==1508==    by 0x1005E4093: readClass(objc_class*, bool, bool) (in /usr/lib/libobjc.A.dylib)
==1508==    by 0x1005E1C13: gc_init (in /usr/lib/libobjc.A.dylib)
==1508==    by 0x1005E924E: objc_initializeClassPair_internal(objc_class*, char const*, objc_class*, objc_class*) (in /usr/lib/libobjc.A.dylib)
==1508==    by 0x1005F6132: layout_string_create (in /usr/lib/libobjc.A.dylib)
==1508==    by 0x1005E483C: realizeClass(objc_class*) (in /usr/lib/libobjc.A.dylib)
==1508==    by 0x1005E4300: copySwiftV1MangledName(char const*, bool) (in /usr/lib/libobjc.A.dylib)
==1508==    by 0x1005E42E9: copySwiftV1MangledName(char const*, bool) (in /usr/lib/libobjc.A.dylib)
==1508==    by 0x1005E42E9: copySwiftV1MangledName(char const*, bool) (in /usr/lib/libobjc.A.dylib)
==1508==    by 0x1005E42E9: copySwiftV1MangledName(char const*, bool) (in /usr/lib/libobjc.A.dylib)
==1508== 
==1508== LEAK SUMMARY:
==1508==    definitely lost: 288 bytes in 2 blocks
==1508==    indirectly lost: 0 bytes in 0 blocks
==1508==      possibly lost: 2,064 bytes in 1 blocks
==1508==    still reachable: 4,096 bytes in 1 blocks
==1508==         suppressed: 20,477 bytes in 189 blocks
```